Lib/doctest.py

```diff
@@ -1044,6 +1045,13 @@ def _find(self, tests, obj, name, module, source_lines, globs, seen):
                 if isinstance(val, (staticmethod, classmethod)):
                     val = val.__func__
 
+                # For functools.cached_property objects, val is considered
+                # to be defined in the functools module instead of the
+                # module being inspected by doctest, so we grab the underlying
+                # function instead to inspect it for doctests.
+                if isinstance(val, functools.cached_property):
+                    val = val.func
+
```

Comment:
I don't love the idea of special-casing `cached_property` in `doctest` -- why should it be treated differently to any other custom descriptor?

This would also fix the issue, but still doesn't feel like a fully general solution (`doctest` will still fail to find the doctests of other custom descriptors if their `__module__` attribute is set to the module in which the custom descriptor class is defined):

```diff
diff --git a/Lib/functools.py b/Lib/functools.py
index be44ccdae6..a2fc28779d 100644
--- a/Lib/functools.py
+++ b/Lib/functools.py
@@ -984,6 +984,7 @@ def __init__(self, func):
         self.func = func
         self.attrname = None
         self.__doc__ = func.__doc__
+        self.__module__ = func.__module__

     def __set_name__(self, owner, name):
         if self.attrname is None:
```

It seems correct for doctest to try to only run doctests that are actually from the module being doctested, though.

I'm curious if @carljm has any thoughts here!

Reply:
I'm not sure there's an ideal solution here. A few notes/thoughts:

1. The existing support for `property` here is pretty ugly; there's just a special-case in `_from_module` so that every property is always considered a part of any module. It seems like if `property` is going to be special-cased it'd be possible to separately check `fget`, `fset` etc for correct module and for docstring; not sure why that wasn't done. I think there's arguably even a bug hiding here: if you have a property with both a getter and a setter, only the getter's docstring is checked (since that's the one that makes it into the `__doc__` of the property itself.) Any doctests in the setter docstring will be silently ignored, I think.
2. If we're going to add support for custom descriptors, I prefer @AlexWaygood 's approach of setting `__module__` on each cached property instance over the approach of this PR. It still doesn't support custom descriptors out of the box (not sure there's any reasonable way to do that), but it at least follows a standard "protocol" that any custom descriptor can participate in if they choose to.

Reply:
I like the sound of solution 1 better. Do you see some downside to it besides preserving special-casing around `property`?

Reply:
Actually, `cached_property` doesn't have an `fget`. Instead the function is held in the attribute `func`. So likely the code would still have to special case `cached_property`. Is there a better way to go about this?

Assuming not, solution 2 sounds more appealing (a more innocent one line change 😁).

Reply:
I don't see a big problem in special-casing some popular stdlib descriptor-like things. When we special-case things, we can be sure they work correctly and do not cause any regressions in other parts of the world.

`functools.cached_property` is clearly a popular stdlib descriptor-like thing :)

Reply:
> I don't see a big problem in special-casing some popular stdlib descriptor-like things.

It's not very principled; it  doesn't scale very well (why should we add special-casing for `cached_property` but not for `types.DynamicClassAttribute` or for `functools.singledispatchmethod`? Why should we privilege classes inside the stdlib over ones in user code? That's not the way we generally do things in the standard library), and it introduces a new dependency between stdlib modules. It's fine to add a new dependency between stdlib modules if there's a good reason, but we should _generally_ try to avoid it where possible given that the stdlib is already a massive import cycle.

> When we special-case things, we can be sure they work correctly and do not cause any regressions in other parts of the world.

What kind of code do you think might be broken if we set `__module__` on each `cached_property` instance @sobolevn? :)

Reply:
> I like the sound of solution 1 better. Do you see some downside to it besides preserving special-casing around `property`?

@tjsmart, I don't think @carljm's points (1) and (2) were meant as alternative solutions to the problem at hand here. I think Carl was just looking at how doctest handles `builtins.property` doctests, and commenting on why that approach _wouldn't_ work for `cached_property`.

Reply:
Thanks for the clarification @AlexWaygood.

Perhaps `cached_property` should be utilizing [`functools.update_wrapper`](https://docs.python.org/3/library/functools.html#functools.update_wrapper)?
> The main intended use for this function is in [decorator](https://docs.python.org/3/glossary.html#term-decorator) functions which wrap the decorated function and return the wrapper. If the wrapper function is not updated, the metadata of the returned function will reflect the wrapper definition rather than the original function definition, which is typically less than helpful.

```diff
diff --git a/Lib/functools.py b/Lib/functools.py
index be44ccdae6..5cbb958ad7 100644
--- a/Lib/functools.py
+++ b/Lib/functools.py
@@ -983,7 +983,7 @@ class cached_property:
     def __init__(self, func):
         self.func = func
         self.attrname = None
-        self.__doc__ = func.__doc__
+        update_wrapper(self, func)
```

Reply:
> Perhaps `cached_property` should be utilizing [`functools.update_wrapper`](https://docs.python.org/3/library/functools.html#functools.update_wrapper)?

That would be a slightly more invasive change than what's strictly required here (we only _need_ `__module__` copied across). Also, not all of the attributes would make sense to be copied across:

```pycon
>>> from functools import cached_property
>>> class Foo:
...     def __init__(self, attr):
...         self.attr = attr
...     has_numeric_attr = cached_property(lambda self: isinstance(self.attr, (int, float)))
...
>>> Foo.has_numeric_attr.attrname
'has_numeric_attr'
>>> Foo.has_numeric_attr.__name__
'<lambda>'
```

If you wanted to do it in one line, I suppose you could do something like this:

```diff
--- a/Lib/functools.py
+++ b/Lib/functools.py
@@ -983,7 +983,7 @@ class cached_property:
     def __init__(self, func):
         self.func = func
         self.attrname = None
-        self.__doc__ = func.__doc__
+        update_wrapper(self, func, assigned=("__module__", "__doc__"))

     def __set_name__(self, owner, name):
         if self.attrname is None:
```

But I'd find it more readable if you just did this:

```diff
--- a/Lib/functools.py
+++ b/Lib/functools.py
@@ -983,7 +983,7 @@ class cached_property:
     def __init__(self, func):
         self.func = func
         self.attrname = None
         self.__doc__ = func.__doc__
+        self.__module__ = func.__module__

     def __set_name__(self, owner, name):
         if self.attrname is None:
```

Reply:
I've updated the PR to follow your solution @AlexWaygood, since that seems to be where this discussion is settling. But I'm still open to further feedback.